### PR TITLE
[eval] save eventstream & llm completions for SWE-Bench run_infer

### PR DESCRIPTION
**Pull request by xingyaoww**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

- Save the actual eventstream history for SWE-Bench run_infer to be consistent with actual events
- Allow saving `llm_completions` to `State` (for saving exact trajectories for future training)

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by enyst** [comment]
I was just thinking of this, we used to be able to configure 
```
file_store="local"
file_store_path="/Users/enyst/repos/odie/cache"
```
to save all events. Each event stream had the instance name. But lately it no longer works when running swe-bench, although it works in all other ways, when `openhands` runs in cli, and with UI. It must be because we now create a new [AppConfig](https://github.com/All-Hands-AI/OpenHands/blob/f7ebc1cf1f1f4ae306d1e630ac57b618d0d85810/evaluation/swe_bench/run_infer.py#L120) instance, which doesn't read config.toml.

Maybe we should still be able to do that, it's very helpful. 🤔

**Comment by xingyaoww** [comment]
@enyst I actually just pushed a commit that try to saves the raw `llm_completions` for each run (when a boolean var is true). I think saving this metadata is definitely helpful for future - but i think reading config from `config.toml` / env var is a little bit too dangerous for eval purposes: e.g., you may unintentionally have a variable there that tweaked an eval setting that makes the score not comparable (e.g., run as openhands, where default eval run as root).

**Comment by enyst** [comment]
You're right, that makes sense, I have a `config.toml.swe`, a `config.toml.integration` and a `config.toml.bak`, because... yep! the wrong vars for the purpose. 😅 

We don't have to use `config.toml` file to take advantage of this filestore feature though: we can add it instead to the AppConfig as we did with the other intended settings. The location is configurable the same way. We already configure an evals-only log dir in the desired place per instance as `infer_logs`, where we save logs, do you see a problem if have another for the event stream?

Would something like this work:

```
config = AppConfig(
        default_agent=metadata.agent_class,
        run_as_openhands=False,
        max_budget_per_task=4,
        max_iterations=metadata.max_iterations,
        runtime=os.environ.get('RUNTIME', 'eventstream'),
        file_store='local'
        file_store_path='.../instance_name/eventstream'
```

> I'm still figuring out what would be the best way to save these trajectories - but i think keeping them in `State` and only written to file if needed (like in an eval script) is probably a good default behavior?

Well, technically histories not saved in State anymore 😅. They're saved in the event stream only. They can still be accessed via `state.history`, which acts as legacy intermediary, but I'm removing this intermediation [in this PR](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-6acd24038b8b1bbaaa864f2dbd0aa64151d17addfb5db45d6685c28d36a803d3R310). There's a single source of truth, and in my understanding, this is the intended design.

**Comment by xingyaoww** [comment]
@enyst I think @tofarr is working on event stream serialization (probably finished in the coming weeks). Before that, i think it is probably ok for us to get the history still from `state.history`? I think the reason i want to put it in `State` is bc it will be saved to `output.jsonl` with other stuff (more convenient than have to go to each specific folder).

**Comment by xingyaoww** [comment]
> Well, technically histories not saved in State anymore 😅. They're saved in the event stream only. They can still be accessed via state.history, which acts as legacy intermediary, but I'm removing this intermediation [in this PR](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-6acd24038b8b1bbaaa864f2dbd0aa64151d17addfb5db45d6685c28d36a803d3R310)

This will be slightly tricky for these eval programs that hope to get the history from the final output of `run_controller` (i.e., `State`). Any idea there? I'm generally against the idea of introducing unexpected side-effect (e.g., having eventstream saved to a random folder), at least for eval script, and hoping to get whatever i needed from the output `State`.

How about we add these to `extra_data` field in `State` which is reserved for eval?

File: evaluation/swe_bench/run_infer.py [hunk 1]
```diff
@@ -383,10 +384,7 @@ def process_instance(
     if state is None:
         raise ValueError('State should not be None.')
 
-    # history is now available as a stream of events, rather than list of pairs of (Action, Observation)
-    # for compatibility with the existing output format, we can remake the pairs here
-    # remove when it becomes unnecessary
-    histories = state.history.compatibility_for_eval_history_pairs()
+    histories = [event_to_dict(event) for event in state.history.get_events()]
```
Comment (enyst) [inline evaluation/swe_bench/run_infer.py:387]: Please note that this method filters out these types of events currently https://github.com/All-Hands-AI/OpenHands/blob/c0201d92f86133259c9bb685b360c14460b8ff02/openhands/memory/history.py#L31

Basically the reason for the existence of the compatibility function is to bring back NullActions and NullObservations when they're necessary for the expected format (action, obs) to be still respected.

Comment (xingyaoww) [reply evaluation/swe_bench/run_infer.py:387]: yeah i think this is actually good for eval purpose - only saving stuff that's needed?

Comment (enyst) [reply evaluation/swe_bench/run_infer.py:387]: IMHO it's good, but it doesn't always match the pattern. Just wondering, will that work fine with the visualizer? Should it be done in all evals eventually?

Comment (xingyaoww) [reply evaluation/swe_bench/run_infer.py:387]: haven't fixed the visualizer for this yet - but eventually i'd actually hope to re-use the Share-OD visualizer for all eval

**Comment by enyst** [comment]
About histories:

> > Well, technically histories not saved in State anymore 😅. They're saved in the event stream only. They can still be accessed via state.history, which acts as legacy intermediary, but I'm removing this intermediation [in this PR](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-6acd24038b8b1bbaaa864f2dbd0aa64151d17addfb5db45d6685c28d36a803d3R310)
> 
> This will be slightly tricky for these eval programs that hope to get the history from the final output of `run_controller` (i.e., `State`). Any idea there? I'm generally against the idea of introducing unexpected side-effect (e.g., having eventstream saved to a random folder), at least for eval script, and hoping to get whatever i needed from the output `State`.
> 
> How about we add these to `extra_data` field in `State` which is reserved for eval?

I see the need, the interesting part, it seems to me, is that this is a question for everyone that currently uses `state.history`, not only evals: in particular, once `state.history` is gone, agents need a reference to the event stream directly or via State or some other intermediary.

I wasn't sure about this, but according to @neubig here, it's fine for agents to have a reference. I'm thinking to try it next on PR https://github.com/All-Hands-AI/OpenHands/pull/3808 

Maybe via State, and then everyone using today `state.history` will just use `state.eventstream`. Though I tried it in the past and gave it up, I don't remember why. 😅  What do you think about it? Cc: @rbren

**Comment by enyst** [comment]
> @enyst I think @tofarr is working on event stream serialization (probably finished in the coming weeks). Before that, i think it is probably ok for us to get the history still from `state.history` similar to what we were doing before?

That sounds interesting, is this work or spec available, or an issue about it? We serialize the event stream, in-memory by default, local or S3 in .json files. Curious where that is heading. 😄

**Review by neubig — APPROVED** [review]
LGTM!

**Pull request merged**